src/qttools/kernels/cuda/banded.py

```diff
@@ -0,0 +1,1071 @@
+        # `b_ptrs` is a block of [BLOCK_SIZE_K, BLOCK_SIZE_N] pointers
+        # See above `Pointer Arithmetic` section for details
+        offs_am = (denseC_row * BLK_M + tl.arange(0, BLK_M)) % M
+        offs_bn = (denseC_col * BLK_N + tl.arange(0, BLK_N)) % M
+        offs_k = tl.arange(0, BLK_K)
+        a_ptrs = a_ptr + (offs_am[:, None] * stride_am + offs_k[None, :] * stride_ak)
+        b_ptrs = b_ptr + (offs_k[:, None] * stride_bk + offs_bn[None, :] * stride_bn)
+
+        if denseC_col >= num_denseC_blocks:
+            return
+        if denseC_row >= num_denseC_blocks:
+            return
+        if denseC_col < 0:
+            return
+        if denseC_row < 0:
+            return
+
+        # # move the pointers to the start of the band
+        a_ptrs += bandA_col_start * BLK_N * stride_ak
+        b_ptrs += bandB_row_start * BLK_M * stride_bk
+
+        if variant == 0:
+            src_dtype = tl.float16
+            dest_dtype = tl.float16
+        elif variant == 1:
+            src_dtype = tl.float32
+            dest_dtype = tl.float32
+        elif variant == 2:
+            src_dtype = tl.float32
+            dest_dtype = tl.float16
+        elif variant == 3:
+            src_dtype = tl.float16
+            dest_dtype = tl.float32
+        elif variant == 4:
+            src_dtype = tl.float64
+            dest_dtype = tl.float32
+        elif variant == 5:
+            src_dtype = tl.float64
+            dest_dtype = tl.float16
+        elif variant == 6:
+            src_dtype = tl.float64
+            dest_dtype = tl.float64
+        # -----------------------------------------------------------
+        # Iterate to compute a block of the C matrix.
+        # We accumulate into a `[BLOCK_SIZE_M, BLOCK_SIZE_N]` block
+        # of fp32 values for higher accuracy.
+        # `accumulator` will be converted back to fp16 after the loop.
+        accumulator = tl.zeros((BLK_M, BLK_N), dtype=tl.float32)
+
+        if variant > 1:
+            for k in range(0, tl.cdiv(num_blocks * BLK_M, BLK_K)):
+                # Load the next block of A and B, generate a mask by checking the K dimension.
+                # If it is out of bounds, set it to 0.
+                a = tl.load(a_ptrs).to(
+                    dest_dtype
+                )  # , mask=offs_k[None, :] < band_a - k * BLK_K, other=0.0)
+                b = tl.load(b_ptrs).to(
+                    dest_dtype
+                )  # , mask=offs_k[:, None] < band_b - k * BLK_K, other=0.0)
+                # We accumulate along the K dimension.
+                accumulator = tl.dot(a, b, accumulator, allow_tf32=allow_tf32)
+                # accumulator += b
+                # Advance the ptrs to the next K block.
+                a_ptrs += BLK_K * stride_ak
+                b_ptrs += BLK_K * stride_bk
+            c = accumulator.to(tl.float16)
+        else:
+            for k in range(tl.cdiv(num_blocks * BLK_M, BLK_K)):
```

Comment:
What is the actual difference with the first branch? I see two differences:
- There is no conversion to the destination datatype. If source and destination datatypes match, is there any actual overhead? If yes, then shouldn't all such variants use the second branch?
- There is no option to `allow_tf32`. I understand this flag can also affect float16 tensor cores (faster and less precise).

Btw, is there an option to target bfloat16?

Reply:
Minor correction: I just saw that `allow_tf32` is not an exact match between PyTorch and Triton. Furthermore, Triton is deprecating `allow_tf32` in favor of `input_precision` (see https://triton-lang.org/main/python-api/generated/triton.language.dot.html). Please leave a TODO to update the call accordingly.